ols/src/llms/providers/azure_openai.py

```diff
@@ -54,24 +68,37 @@ def default_params(self) -> dict[str, Any]:
             "verbose": False,
         }
 
-        # TODO: OLS-679 Azure AD token can be reused w/o the need to refresh it for each request
         if self.credentials is not None:
             # if credentials with API key is set, use it to call Azure OpenAI endpoints
             default_parameters["api_key"] = self.credentials
         else:
-            # credentials for API key is not set
-            # -> it is needed to retrieve token first
-            access_token = self.retrieve_access_token(azure_config)
-            # set up active directory token to access Azure services, including OpenAI one
-            if access_token is not None:
-                default_parameters["azure_ad_token"] = access_token.token
-
+            # credentials for API key is not set -> azure AD token is
+            # obtained through azure config parameters (tenant_id,
+            # client_id and client_secret)
+            access_token = self.resolve_access_token(azure_config)
+            default_parameters["azure_ad_token"] = access_token
         return default_parameters
 
     def load(self) -> LLM:
         """Load LLM."""
         return AzureChatOpenAI(**self.params)
 
+    def resolve_access_token(self, azure_config: AzureOpenAIConfig) -> Optional[str]:
+        """Resolve access token to call Azure OpenAI."""
+        if token_is_expired():
+            access_token = self.retrieve_access_token(azure_config)
+            # set up active directory token to access Azure services, including OpenAI one
+            if access_token is None:
+                # we are unable to retrieve access token
+                token_cache["access_token"] = access_token  # None
+            else:
+                token_cache["access_token"] = access_token.token
+                # Update expires_on to current time + expires_in,
+                # minus a small leeway. Default expiration seems to
+                # be 1 hour.
+                token_cache["expires_on"] = time.time() + access_token.expires_on - 30
```

Comment:
will it work in non-UTC environment?

Reply:
is the `expires_on` some relative time (duration)?

Reply:
It is an int - `The token's expiration time in Unix time.`
I should probably explicitly make timestamps UTC.